File: app/src/main/java/me/tagavari/airmessage/helper/ConnectionServiceLaunchHelper.java
package me.tagavari.airmessage.helper;

import android.content.Context;
import android.content.Intent;

import androidx.core.content.ContextCompat;

import me.tagavari.airmessage.data.SharedPreferencesManager;
import me.tagavari.airmessage.enums.ProxyType;
import me.tagavari.airmessage.service.ConnectionService;

public class ConnectionServiceLaunchHelper {
	/**
	 * Launches the connection service a either a temporary or persistent foreground service, depending on the selected proxy type
	 */
	public static void launchAutomatic(Context context) {
		switch(SharedPreferencesManager.getProxyType(context)) {
			case ProxyType.direct:
				launchPersistent(context);
				break;
			case ProxyType.connect:
				launchTemporary(context);
				break;
		}
	}
	
	/**
	 * Launches the connection service as a persistent foreground service
	 */
	public static void launchPersistent(Context context) {
		ContextCompat.startForegroundService(context,
				new Intent(context, ConnectionService.class)
						.setAction(ConnectionService.selfIntentActionConnect)
						.putExtra(ConnectionService.selfIntentExtraForeground, true)
		);
	}
	
	/**
	 * Launches the connection service as a temporary foreground service
	 */
	public static void launchTemporary(Context context) {
		ContextCompat.startForegroundService(context,
				new Intent(context, ConnectionService.class)
						.setAction(ConnectionService.selfIntentActionConnect)
						.putExtra(ConnectionService.selfIntentExtraTemporary, true)
						.putExtra(ConnectionService.selfIntentExtraForeground, true)
		);
	}
}

File: common/me/tagavari/airmessage/common/helper/ConnectionServiceLaunchHelper.kt
package me.tagavari.airmessage.common.helper

import android.content.Context
import android.content.Intent
import androidx.core.content.ContextCompat
import me.tagavari.airmessage.common.data.SharedPreferencesManager
import me.tagavari.airmessage.common.enums.ProxyType
import me.tagavari.airmessage.service.ConnectionService

object ConnectionServiceLaunchHelper {
	/**
	 * Launches the connection service a either a temporary or persistent foreground service, depending on the selected proxy type
	 */
	@JvmStatic
	fun launchAutomatic(context: Context) {
		when(SharedPreferencesManager.getProxyType(context)) {
			ProxyType.direct -> launchPersistent(context)
			ProxyType.connect -> launchTemporary(context)
		}
	}
	
	/**
	 * Launches the connection service as a persistent foreground service
	 */
	@JvmStatic
	fun launchPersistent(context: Context) {
		ContextCompat.startForegroundService(context,
				Intent(context, ConnectionService::class.java).apply {
					action = ConnectionService.selfIntentActionConnect
					putExtra(ConnectionService.selfIntentExtraForeground, true)
				}
		)
	}
	
	/**
	 * Launches the connection service as a temporary foreground service
	 */
	@JvmStatic
	fun launchTemporary(context: Context) {
		ContextCompat.startForegroundService(context,
				Intent(context, ConnectionService::class.java).apply {
					action = ConnectionService.selfIntentActionConnect
					putExtra(ConnectionService.selfIntentExtraTemporary, true)
					putExtra(ConnectionService.selfIntentExtraForeground, true)
				}
		)
	}
}